Enhancement: Add New SLA Types (#16902)
PrefectHQ/prefect: docs/v3/automate/events/slas.mdx

deployments:
  my-deployment:
    sla:
        - name: "time-to-completion"
          duration: 10
          severity: "high"
        - name: "lateness"
          within: 600
          severity: "high"
```
</Accordion>

<Accordion title="Defining SLAs using a .deploy method">
```python .deploy SLA
    from prefect import flow
    from prefect._experimental.sla.objects import TimeToCompletionSla, LatenessSla

    @flow
    def my_flow():
        pass

    flow.from_source(
        source=source,
        entrypoint="my_file.py:my_flow",
    ).deploy(

>         name="my deployment",

Reviewer: This is legal, but forces you to use quotes around the fully qualified deployment name (e.g. `prefect deployment run "my_flow/my deployment"`). The following is slightly more forgiving:

```suggestion
        name="my-deployment",
```

        work_pool_name="my_pool",
        _sla=[
            TimeToCompletionSla(name="my-time-to-completion-sla", duration=10, severity="high"),
            LatenessSla(name="my-lateness-sla", within=timedelta(minutes=10), severity="high"),

>         ]

Reviewer: Can you include an example of a Frequency SLA here too?

    )
```
</Accordion>    

<Accordion title="Defining SLAs with the Prefect CLI">
```bash CLI SLA
prefect deploy --sla '{"name": "my-time-to-completion-sla", "duration": 10, "severity": "high"}'
```
</Accordion>
</AccordionGroup>

## Types of SLAs

### Time to Completion SLA

A Time to Completion SLA describes how long flow runs should take and how serious running over that time is. The SLA is triggered when a flow run takes longer than the specified duration to complete. 

The Time to Completion SLA requires one unique parameter:
- `duration`: The maximum allowed duration in seconds before the SLA is violated

This SLA is useful for monitoring long-running flows and ensuring they complete within an expected timeframe. For example, if you set a Time to Completion SLA on your deployment with a `duration` of 600 seconds (10 minutes) the backend will emit an `prefect.sla.violation` event if a flow run takes longer than 10 minutes to complete.

### Frequency SLA

A Frequency SLA describes the interval a deployment should run at and how serious missing that interval is.

The Frequency SLA requires one unique parameter:

> - `stale_after`: The maximum allowed time between flow runs (e.g. 1 hour for hourly jobs)

Reviewer: Is `1 hour` a valid setting for this parameter? If so, I suggest making it a code literal.

```suggestion
- `stale_after`: The maximum allowed time between flow runs (e.g. `1 hour` for hourly jobs)
```

> For example, if you have a deployment that should run every hour, you can set up a Frequency SLA with a `stale_after` of 1 hour - this will trigger if more than an hour passes between `Completed` flow runs.

Reviewer: ```suggestion
For example, if you have a deployment that should run every hour, you can set up a Frequency SLA with a `stale_after` of 1 hour.
This SLA triggers when more than an hour passes between `Completed` flow runs.
```

### Lateness SLA

> A Lateness SLA describes how serious it is when a flow run fails to start at its scheduled time.

Reviewer: ```suggestion
A Lateness SLA describes the severity of a flow run failing to start at its scheduled time.
```

The Lateness SLA requires one unique parameter:

> - `within`: The time within the previous flow run before the SLA is violated
> 
> For example, if a deployment with a Lateness SLA with a `within` of 5 minutes is scheduled to run every day at 10am, and a flow run is not started by 10:05am, the backend will emit an SLA violation event.

Reviewer: I tried wordsmithing this a bit.

```suggestion
- `within`: The amount of startup time allotted to a flow run before the SLA is violated.

For example, if a deployment is scheduled to run every day at 10am, and has a Lateness SLA of 5 minutes, the backend will emit an SLA violation event if a flow run does not start by 10:05am.
```
